flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/stream/StreamExecMultiJoin.java

```diff
@@ -0,0 +1,444 @@
+        final JoinCondition[] instantiatedJoinConditions = new JoinCondition[joinConditions.size()];
+        for (int i = 0; i < joinConditions.size(); i++) {
+            RexNode rexCond = joinConditions.get(i);
+            if (rexCond != null) {
+                GeneratedJoinCondition generatedCondition =
+                        generateJoinConditionForInput(
+                                config, classLoader, rexCond, inputRowTypes, i);
+                try {
+                    assert generatedCondition != null;
+                    instantiatedJoinConditions[i] = generatedCondition.newInstance(classLoader);
+                } catch (Exception e) {
+                    throw new RuntimeException(
+                            "Failed to instantiate join condition for input " + i, e);
+                }
+            } else {
+                instantiatedJoinConditions[i] = null;
+            }
+        }
+        return instantiatedJoinConditions;
+    }
+
+    private StreamOperatorFactory<RowData> createOperatorFactory(
+            ExecNodeConfig config,
+            List<InternalTypeInfo<RowData>> inputTypeInfos,
+            JoinCondition[] joinConditions,
+            JoinKeyExtractor keyExtractor) {
+        List<Long> stateTtls =
+                StateMetadata.getStateTtlForMultiInputOperator(
+                        config, inputTypeInfos.size(), stateMetadataList);
+        long[] stateRetentionTimes = stateTtls.stream().mapToLong(Long::longValue).toArray();
+
+        return new StreamingMultiJoinOperatorFactoryImpl(
+                inputTypeInfos,
+                inputSideSpecs,
+                joinTypes,
+                null, // multiJoinCondition is currently not used
+                stateRetentionTimes,
+                joinConditions,
+                keyExtractor,
+                joinAttributeMap);
+    }
+
+    private Transformation<RowData> createTransformation(
+            List<Transformation<RowData>> inputTransforms,
+            TransformationMetadata metadata,
+            StreamOperatorFactory<RowData> operatorFactory) {
+        if (inputTransforms.isEmpty()) {
+            throw new IllegalStateException("StreamExecMultiJoin requires at least two inputs.");
+        }
+
+        return ExecNodeUtil.createKeyedMultiInputTransformation(
+                inputTransforms,
+                commonJoinKeySelectorsForTransformation,
+                commonJoinKeyType,
+                metadata,
+                operatorFactory,
+                InternalTypeInfo.of(getOutputType()),
+                inputTransforms.get(0).getParallelism(),
+                false);
+    }
+
+    private GeneratedJoinCondition generateJoinConditionForInput(
+            ExecNodeConfig config,
+            ClassLoader classLoader,
+            RexNode joinCondition,
+            List<RowType> inputRowTypes,
+            int inputIndex) {
+        if (inputIndex == 0) {
```

Comment:
good practices: comments like `// Create transformation` above are kind of unimportant, but here is a typical location that definitely deserves a comment. what does this mean?